feat(): pg-query-stream typescript

chyzwar opened the pull request:
New version should emit types compatible with
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/47e9cf6d8947e46a82bb09bbcc3dd74a3ac0ffa1/types/pg-query-stream/index.d.ts

Let me know if this is going to right direction?

TODO:
- [x] Confirm readable option emitClose, is this is valid option? 
- [x] Add mocha setup for typescript
- [x] Convert tests to typescript
- [x] Ensure that emitted code is compatible with JS version

brianc commented on packages/pg-query-stream/tsconfig.json:
```diff
@@ -0,0 +1,26 @@
+    "module": "commonjs",
+    "esModuleInterop": true,
+    "allowSyntheticDefaultImports": true,
+    "strict": false,
```
it will be nice to turn this to `true` one day, but this is fine for the initial port.

chyzwar replied:
I think this would be possible after pg-cursor is converted. I will look at pg-cursor next

brianc reviewed (commented):
this looks rad - lmk if you have any questions!

chyzwar commented:
I think this should be ready to review.

- Added solution project to compile pg-query-stream and pg-protocol (incremental and composite)
- Added files in pg-query-stream and pg-protocol to avoid publishing test files and buildInfo
- Remove emitClose from super call in QueryStream
- Remove unsued results from _read -> read
- Converted test to ts and changed var top consts
- Rename class from PgQueryStream to QueryStream

brianc commented on packages/pg-query-stream/src/index.ts:
```diff
@@ -1,11 +1,30 @@
+  public constructor(text: string, values?: any[], config: QueryStreamConfig = {}) {
     const { batchSize, highWaterMark = 100 } = config
-    // https://nodejs.org/api/stream.html#stream_new_stream_readable_options
-    super({ objectMode: true, emitClose: true, autoDestroy: true, highWaterMark: batchSize || highWaterMark })
```
curious: why remove `emitClose` here?

brianc commented:
This looks solid - thanks a lot for doing this!  Sorry for the delay....have had a really horrible October - my doggie died. 😭 back in the saddle again now though.

chyzwar commented:
emitClose is true by default and typescript was throwing error when It was trying to pass this option.

chyzwar commented:
Sorry to hear about your dog.

brianc commented:
@chyzwar sorry for the delay...life is crazy.  I'm down to merge this but there are a few merge conflicts.  Care to take a quick look at those & resolve them? Then i'll get this merged!  Thanks again! ❤️

chyzwar commented:
this now has been rebased on master.

brianc commented:
awesome! thanks! Will get this merged up today!

brianc commented:
release pg@8.5.0

vitaly-t commented:
Added issue #2412
